### PR TITLE
Tim's updates, see description (Closes #74 #95 #124 #139 #146 #157 #158)

### LINKED ISSUE #74: Enhancement: regex searches
Could the search box default to regex? Perhaps as a preference setting?

### PR DESCRIPTION
**Pull request by timriker**
regex searching
"Move the content of this folder to the configured report root" -> underlay 
new settings in xml!
added preferences.json, removed defaultpreferences.json

new tab/button styles and icons
expand/collapse/normal toggle on tables
pool, irule and datagroup tables use "details" html5 markup
stats to loggederrors.json
hide main desc, asm, ssl, compression, persistence columns by default
hide certs country, state by default
show snat pool in virtual server details
add description column to pool table
url and pool links in datagroups
new status search syntax
Fix curl headers
add monitor links
monitor names in pooltable
remove hidden textstatus markup
merge code to resetFilters() and toggleExpandCollapseRestore()

### CODE REVIEW

**Comment by epacke** [comment]
Looks good. Just one change left. 

This section:
> "BigIPReport is written and maintained by Patrik Jonsson"
should contain your name.

**Comment by epacke** [comment]
I ran into a bunch of exceptions. Need to troubleshoot a bit.

**Comment by epacke** [comment]
I've tried to find the common denominator, but haven't been able to. Either way, I think it's about time that we started to migrate the script over to REST because it's likely faster, supported by F5 and we can start running the script on linux/containers. :)

**Comment by timriker** [comment]
I fixed an issue dealing with datagroups with name entries without values. Can you test if that is your issue?
I do think we should move to REST. People can use admin accounts if they still have < v12 servers. I have a few that we have not yet turned off, but we are really close. Testing 14 now. Still not doing any active-active, but I hope to test that in our sandbox soon.
Test out the fix? it's this:
`if (!$TempPool) {continue}`

**Comment by epacke** [comment]
Regarding active-active, what changes would you propose? Will test the fix
today.

**Comment by epacke** [comment]
The fix did not work:

> Cannot index into a null array.
> At C:\Scripts\BigipReport\new\bigipreport.ps1:1183 char:13
> +             $Value = [string]$AddressClassValues[$i][$x]
> +             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>     + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
>     + FullyQualifiedErrorId : NullArray

It seems to be thrown globally across some units for **all data group lists** and not thrown **per** data group. I tried rebooting but no luck.

**Comment by timriker** [comment]
That line is with your patch, right? Mine still has:
 `  [array]$AddressClassValues = $F5.LocalLBClass.get_address_class_member_data_value($AddressClassKeys)`
Many of my address datagroups have no comments in the value side. I'll test with a mix and see if I can cause any issues.

**Comment by epacke** [comment]
> That line is with your patch, right? Mine still has:

No, with the latest commit in this PR.

**Comment by timriker** [comment]
I see that line of code here:
https://github.com/epacke/BigIPReport/pull/153/commits/f01867f8442b9e0801294e53cb6a59061a9676b7#diff-621cd034392d21b7a2726b46f96c59e8L1181
Which is your PR that I did not apply.

**Comment by timriker** [comment]
Do you have F5s with no address datagroups? or just one? If one, does it have values? So far, I've not been able to duplicate the error.

**Comment by timriker** [comment]
I wrapped the line in a try catch, which should avoid the issue. Test?

**Comment by epacke** [comment]
> Do you have F5s with no address datagroups? or just one? If one, does it have values? So far, I've not been able to duplicate the error.

Actually, the bug concerns the whole LB no matter if a data group is empty or not.

**Comment by epacke** [comment]
Let's not add any more functionality to this PR if that's OK? Since there's no unit/integration testing it takes a long time to go through everything every time. :)

**Comment by timriker** [comment]
The REST datagroup access is failing on a couple of my boxes. Still testing. Will comment as I learn more.

**Comment by timriker** [comment]
I think this is on my end. One one of our largest boxes, I'm getting a few errors on the irule saving virtualservers. Not sure why. Might be a false match. In order to pull ASM, a Guest or Auditor account is not enough. :( I've upped accounts to Administrator for testing.
I'm fine with this pull request getting approved. I have a few small changes coming, adding regex to site defaults, etc. but it can go before if you like.

**Comment by epacke** [comment]
Hi Tim
Now the code runs without errors.

A few questions/feedback:

> new status search syntax
How does it work?

> do we need textstatus html anymore?
I don't understand this one.

> regex searching
How is it enabled?

> expand/collapse/normal toggle
Could you please print out what the button does instead of using a single letter? I think it would be better from a UX perspective?

> monitor names in pooltable
I don't see the difference here.

The colors of the table seems to have changed from Gray to Blueish purple. Was this change intentional?

File: underlay/js/bigipreport.js [hunk 1]
```diff
@@ -2070,9 +2161,10 @@ function showPreferences(updatehash) {
     //Populate the content
     $("div#preferences").html(settingsContent);
 
-    //Populate the settings according to the local storage or default settings of none exist
+    //Populate the settings according to the local storage or default settings if none exist
     $("#autoExpandPools").prop("checked", localStorage.getItem("autoExpandPools") === "true");
     $("#adcLinks").prop("checked", localStorage.getItem("showAdcLinks") === "true");
+    $("#regexSearch").prop("checked", localStorage.getItem("regexSearch") === "true");
```
Comment (epacke) [inline underlay/js/bigipreport.js]: This checkbox is missing in my version (I think)?

File: underlay/json/defaultpreferences.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
+  "autoExpandPools": false,
```
Comment (epacke) [inline underlay/json/defaultpreferences.json]: I reckon this should probably be generated by the script using the xml config instead. Ideally we should not spread out the config settings if we can avoid it. What do you think?

File: underlay/js/bigipreport.js [hunk 3]
```diff
@@ -283,34 +282,40 @@ function initializeStatusVIPs() {
...
+        return '<span class="statusicon"><img src="images/black-circle-checkmark.png" alt="Available (Disabled)" title="' + mStatus + ' - Member is available, but disabled"/></span><span class="textstatus">' + mStatus + '</span>'
     } else if (mStatus == "DISABLED:BLUE") {
-        return '<span class="statusicon"><img src="images/black-circle-checkmark.png" alt="Unknown (Disabled)" title="Member is disabled"/></span><span class="textstatus">DISABLED</span>';
+        return '<span class="statusicon"><img src="images/black-circle-checkmark.png" alt="Unknown (Disabled)" title="' + mStatus + ' - Member is disabled"/></span><span class="textstatus">' + mStatus + '</span>';
     }
     return mStatus;
 }
 
-function PoolStatus(pool) {
+function PoolStatus(pool, type) {
```
[the remark this answers is not in the document]
Comment (epacke) [reply underlay/js/bigipreport.js]: In the future we should probably have function definitions too. But that's a later "joy". :)

[the remark this answers is not in the document]
Comment (timriker) [reply underlay/js/bigipreport.js]: type is a pass through from the "type" option in datatables.
https://datatables.net/reference/option/columns.render

File: underlay/js/bigipreport.js (hunk 1 above)
Comment (timriker) [reply underlay/js/bigipreport.js]: It's on the preferences page on mine. Label:
"Use Regular Expressions when searching"
Now defaulting to on.

**Comment by timriker** [comment]
new status search syntax - How does it work? - see the help page. I changed the status text. I added hover titles so that the statustext equivalent for an icon is visible.

do we need textstatus html anymore?  - I don't understand this one. - there is a span.textstatus that is not displayed. The new text status is rendered to datatables for the filter view. I think we can remove it from the display and print views. We can already search on it without it being in the html.

regex searching - How is it enabled? - in my pull request it's now enabled by default, and can be turned off under the preferences tab.

expand/collapse/normal toggle - Could you please print out what the button does instead of using a single letter? I think it would be better from a UX perspective? - I was looking for some chevron type indications, but I'll revert to full words for now until we find some other UI way of showing it. I'd like to add the same button to the other table views. Did you see my pull request with the "details/summary" implementation? nice that the browsers now have that built in, not so nice that the "summary" has to stay visible when expanded. 

monitor names in pooltable - I don't see the difference here. - new searchable column, disabled by default. Toggle it on to see. :)

The colors of the table seems to have changed from Gray to Blueish purple. Was this change intentional? - I'm trying to not have the first row the same color as the header. I added a little blue to differentiate them. Any color difference will do, feel free to change. Did you notice the row hover highlight now? or the sort column highlight? Those are included in the datatables "display" class.

File: underlay/json/defaultpreferences.json (hunk 2 above)
Comment (timriker) [reply underlay/json/defaultpreferences.json]: Yes, the default preferences should probably all come from the xml, and either all be in the index.html, or all be written to defaultpreferences.json
We're almost to the point where index.html is a static file. If preferences and the "build date" info were also in preferences json, that might be all that's left?
definedRules is not used anymore, I'll remove that one.
Thoughts?

**Comment by epacke** [comment]
> new status search syntax - How does it work? - see the help page. I changed the status text. I added hover titles so that the statustext equivalent for an icon is visible.

I see it now. Nice!

> do we need textstatus html anymore? - I don't understand this one. - there is a span.textstatus that is not displayed. The new text status is rendered to datatables for the filter view. I think we can remove it from the display and print views. We can already search on it without it being in the html.

If it works without the span, let's remove it. :)

> do we need textstatus html anymore? - I don't understand this one. - there is a span.textstatus that is not displayed. The new text status is rendered to datatables for the filter view. I think we can remove it from the display and print views. We can already search on it without it being in the html. 

I see it now. This is great!

>  I was looking for some chevron type indications, but I'll revert to full words for now until we find some other UI way of showing it. I'd like to add the same button to the other table views. Did you see my pull request with the "details/summary" implementation? nice that the browsers now have that built in, not so nice that the "summary" has to stay visible when expanded.

Good stuff. I will have a look at the other PR soon. :)

> monitor names in pooltable - I don't see the difference here. - new searchable column, disabled by default. Toggle it on to see. :)

I can see this too now. Must have been some cache that expired over the night. Or maybe I need glasses. Nice!

> The colors of the table seems to have changed from Gray to Blueish purple. Was this change intentional? - I'm trying to not have the first row the same color as the header. I added a little blue to differentiate them. Any color difference will do, feel free to change. Did you notice the row hover highlight now? or the sort column highlight? Those are included in the datatables "display" class.

I really like the hover highlights and I agree that the old style looked a bit "dull". However the blue tint looks out of place for me. Any chance this part can be restored together with the borders? I can look into making PR additions in the weekend if you don't have time right now.

**Comment by timriker** [comment]
I reverted headerbackgroundcolor to the previous.
Not sure what border look you are going for. Adding class "display" changed the defaults. Feel free to alter styling. It might be worth removing the collapse entries and testing that with the defaults. datatables has an additional "cell-border" style which I didn't add. I think the current borders and collapse setting are still in place.
I actually like the "compact" style. Considering adding a preferences toggle for that one. :)
https://datatables.net/manual/styling/classes

**Comment by epacke** [comment]
Love the new icons, where did they come from?

It's very important that they're license and free of attribution etc so we don't get into legal trouble.
You mentioned this in the mail and I know this was a problem with some of the buttons before too, but the readability is quite bad when hovering over buttons:
image

I will add this to the back log.

The Close Pool details button seems to have disappeared (same with Virtual server)
image

This should be changed to authors (plural) and maybe point towards the Devcentral thread:
image

You mentioned details section in your mail, what did you mean by that?

If you don't have time/energy to fix any of the above I am OK to merge now. Your call.

**Comment by timriker** [comment]
> Love the new icons, where did they come from?
> 
> It's very important that they're license and free of attribution etc so we don't get into legal trouble.
> 
> You mentioned this in the mail and I know this was a problem with some of the buttons before too, but the readability is quite bad when hovering over buttons:
> image
> 
> I will add this to the back log.
> 
> The Close Pool details button seems to have disappeared (same with Virtual server)
> image
> 
> This should be changed to authors (plural) and maybe point towards the Devcentral thread:
> image
> 
> You mentioned details section in your mail, what did you mean by that?
> 
> If you don't have time/energy to fix any of the above I am OK to merge now. Your call.

My co-worker Jon built the icons. They are built from scratch and under no restrictions. The gear was a royalty free, free use pic originally before he modified it. The F5 logo came from the F5 media kit. I think it's valid use. If someone objects, we remove it.

Pushing fixes for the other issues you raised. Good catch!

**Comment by timriker** [comment]
Alright. ChangeLog updated, mentioned issues fixed. zip distro created. This is ready to merge.

**Comment by epacke** [comment]
Well done Tim!

**Comment by epacke** [comment]
I will update the instructions page and will ping you later if you're willing to give feedback?

**Comment by timriker** [comment]
Great! The help page could use more regex search examples. I added a few. The Folder rename is probably the biggest issue for the instructions. Now called just "underlay" instead of "Move...". The new config xml has more settings in it, so people upgrading will need to add those new settings.

**Comment by epacke** [comment]
I've uploaded the new version as beta on DC (figured we can let the brave people try it for a while before changing to stable), but I feel it would be more suitable for you to present the changes as you've done almost all of the work this release.

**Comment by timriker** [comment]
I commented on the thread:

https://devcentral.f5.com/s/feed/0D51T00006j4YiC

I wish comments were handled better on DC. I didn't see a way to link to my comment, for example.

**Pull request merged**